llsd/serde_xml.py

```diff
@@ -159,42 +191,66 @@ def _indent(self):
         "Write an indentation based on the atom and indentation level."
         self.stream.writelines([self._indent_atom] * self._indent_level)
 
-    def _ARRAY(self, v):
-        "Recursively format an array with pretty turned on."
-        self.stream.write(b'<array>\n')
-        self._indent_level += 1
-        for item in v:
-            self._indent()
-            self._generate(item)
-            self.stream.write(b'\n')
-        self._indent_level -= 1
-        self._indent()
-        self.stream.write(b'</array>')
-
-    def _MAP(self, v):
-        "Recursively format a map with pretty turned on."
-        self.stream.write(b'<map>\n')
-        self._indent_level += 1
-        # sorted list of keys
-        for key in sorted(v):
-            self._indent()
-            self._elt(b'key', UnicodeType(key))
-            self.stream.write(b'\n')
-            self._indent()
-            self._generate(v[key])
-            self.stream.write(b'\n')
-        self._indent_level -= 1
-        self._indent()
-        self.stream.write(b'</map>')
-
     def _write(self, something):
         """
-        Serialize a python object to self.stream as 'pretty' application/llsd+xml.
+        Serialize a python object to self.stream as application/llsd+xml.
+
+        :param something: A python object (typically a dict) to be serialized.
 
-        :param something: a python object (typically a dict) to be serialized.
+        NOTE: This is nearly identical to the above _write with the exception
+        that this one includes newlines and indentation.  Doing something clever
+        for the above may decrease performance for the common case, so it's been
+        split out.  We can probably revisit this, though.
```

Comment:
Comparing the two `_write()` methods, I see only `self._indent()` calls and changes to `self._indent_level` in this one. It's really tempting to:
* give `LLSDXMLFormatter` an `_eol` attribute of `b''`, versus `LLSDXMLPrettyFormatter`'s `_eol` of `b'\n'`
* give `_indent()` an int parameter, default 0
* introduce an empty `LLSDXMLFormatter._indent()` method
* make `LLSDXMLPrettyFormatter._indent()` add a negative param to `self._indent_level` before the `writelines()` call
* make it add a non-negative param to `self._indent_level` after the `writelines()` call
* call `self._indent(1)` for `_MAP` and `_ARRAY`
* call `self._indent(-1)` on `StopIteration`
* add `self._eol` to various of the `write()` calls

and see what the performance hit really is.